# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [4]:
import pandas as pd
import sqlite3
import warnings
warnings.simplefilter('ignore')

C:\Users\AdvancedGaming\AppData\Local\Temp\ipykernel_2520\955563736.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [5]:
con_sales = sqlite3.connect("go_sales.sqlite")
con_sales

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [6]:
sql_query = "SELECT name FROM sqlite_master WHERE name='product' OR  name='product_type' OR  name='product_line' OR  name='sales_staff' OR  name='sales_branch' OR  name='retailer_site' OR  name='country' OR  name='order_header' OR  name='order_details' OR  name='SALES_TARGETData' OR  name='returned_item' OR  name='return_reason';"
pd.read_sql(sql_query, con_sales)

name
0            country
1      order_details
2       order_header
3            product
4       product_line
5       product_type
6      retailer_site
7      return_reason
8      returned_item
9       sales_branch
10       sales_staff
11  SALES_TARGETData

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [7]:
product = pd.DataFrame(pd.read_sql_query('SELECT * from product', con_sales))
product['PRODUCTION_COST'] = product['PRODUCTION_COST'].astype(float)
product100 = product["PRODUCTION_COST"] < 100
product50 = product["PRODUCTION_COST"] > 50
filtered_product = product100 & product50
product.loc[(filtered_product), ('PRODUCTION_COST', 'PRODUCT_NAME')]

PRODUCTION_COST                     PRODUCT_NAME
13            60.00                Blue Steel Putter
14            81.80            Blue Steel Max Putter
17            80.00              Course Pro Golf Bag
24            60.00                  Hibernator Lite
25            86.00                       Hibernator
28            78.55  Hibernator Self - Inflating Mat
31            65.33              Hibernator Camp Cot
32            62.50     Canyon Mule Climber Backpack
56            52.86          Granite Climbing Helmet
58            55.00            Husky Harness Extreme
69            57.13                   Granite Hammer
71            85.11        TrailChef Deluxe Cook Set
81            73.33                        Polar Ice
83            80.00                     Polar Sports
84            60.00                       Polar Wave
85            72.50                    Polar Extreme
88            80.00                     Edge Extreme
91            79.19                        Seeker 35
92            92.58                        Seeker 50
93            75.00           TrailChef Double Flame
94            94.12                   Seeker Extreme
97            56.25                   Glacier Deluxe
98            78.55                      Glacier GPS

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [8]:
margin = pd.DataFrame(pd.read_sql_query('SELECT product_name, margin from product', con_sales))
margin['MARGIN'] = margin['MARGIN'].astype(float)
margin60 = margin["MARGIN"] < 0.2
margin20 = margin["MARGIN"] > 0.6
filtered_margin = margin20 | margin60
margin.loc[(filtered_margin), :]

PRODUCT_NAME  MARGIN
29                 Hibernator Pad    0.17
32   Canyon Mule Climber Backpack    0.17
100             BugShield Natural    0.67
101               BugShield Spray    0.67
102         BugShield Lotion Lite    0.70
103              BugShield Lotion    0.63
105             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [9]:
country = pd.read_sql_query('SELECT * from country', con_sales)
filtered_results = country["CURRENCY_NAME"] == "francs"
country_df = pd.DataFrame(country.loc[filtered_results, 'COUNTRY'])
country_df


COUNTRY
0        France
7       Belgium
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [10]:
product
product["MARGIN"] = product["MARGIN"].astype(float)
filtered_results = product["MARGIN"] > 0.5
product.loc[filtered_results, ('INTRODUCTION_DATE', 'MARGIN')]


INTRODUCTION_DATE  MARGIN
4          10-12-2019    0.51
6          18-12-2019    0.51
7          27-12-2019    0.55
9           13-1-2020    0.55
14         15-12-2019    0.55
17         15-12-2019    0.60
79         26-10-2016    0.51
100         15-2-2011    0.67
101         15-2-2011    0.67
102         15-2-2011    0.70
103         15-2-2011    0.63
105         15-2-2011    0.63
106         15-2-2011    0.60
107         15-2-2011    0.60
108         15-2-2011    0.60
109         15-2-2011    0.60
114         15-2-2011    0.60

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [11]:
sales_branch = pd.read_sql_query('SELECT * from sales_branch', con_sales)
filtered_branches = (sales_branch["ADDRESS2"].notna()) & (sales_branch["REGION"].notna())
sales_branch.loc[filtered_branches, ('ADDRESS2', 'REGION')]

ADDRESS2         REGION
10  Lomas de Santa Fe         México
11        Shinjuku-ku          Tokyo
12           Asahi-ku          Osaka
13            Level 2            VIC
15   CJ 231 2o. Andar             SP
21      Market Street  West Midlands
27      4e verdieping  Noord-Holland

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [12]:
country = pd.read_sql_query('SELECT * from country', con_sales)
filtered_results = (country["CURRENCY_NAME"] == "dollars") | (country["CURRENCY_NAME"] == "new dollar")
country_df = pd.DataFrame(country.loc[filtered_results, 'COUNTRY'])
country_df


COUNTRY
3          Taiwan
6       Australia
14  United States
15         Canada

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [13]:
site = pd.read_sql("SELECT * FROM retailer_site", con_sales)
filtered_branches = (site["POSTAL_ZONE"].str[0] == 'D') & (site["ADDRESS2"].notna()) 
site.loc[filtered_branches, ("ADDRESS1", "ADDRESS2", "CITY")]


ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [14]:
retoeren = pd.read_sql_query("SELECT * FROM returned_item", con_sales)
retoeren["RETURN_QUANTITY"] = retoeren["RETURN_QUANTITY"].astype(float)
retoeren["RETURN_QUANTITY"].astype(float).sum()


14306.0

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [15]:
sales = sales_branch["REGION"].dropna() 
sales.count()

15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [16]:
margin_min = product["MARGIN"].min()
margin_max = product["MARGIN"].max()
margin_avg = product["MARGIN"].mean()

mar_dict = {
    "Minimum Margin": margin_min,
    "Maximum Margin": margin_max,
    "Average Margin": margin_avg
}

df = pd.DataFrame(mar_dict, index=[1])

df.round(2)


Minimum Margin  Maximum Margin  Average Margin
1            0.17             0.7             0.4

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [17]:
AantalKlanten = site["ADDRESS2"].isna()
site.loc[AantalKlanten, 'RETAILER_SITE_CODE'].count()

286

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [18]:
order_details = pd.read_sql("SELECT * FROM order_details", con_sales)
order_details["UNIT_SALE_PRICE"] = order_details["UNIT_SALE_PRICE"].astype(float)
order_details["UNIT_PRICE"] = order_details["UNIT_PRICE"].astype(float)
filtered_units = order_details["UNIT_SALE_PRICE"] < order_details["UNIT_PRICE"]
filtered_table = order_details.loc[filtered_units, "UNIT_COST"]
filtered_table = filtered_table.astype(float)
mean_unit_cost = filtered_table.mean()
mean_unit_cost

125.42305809556548

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [19]:
sales_staff = pd.read_sql("SELECT * FROM sales_staff", con_sales)
sales_staff_temp1 = sales_staff.groupby("POSITION_EN", as_index=False)["FIRST_NAME"].count()
sales_staff_temp = sales_staff_temp1.rename(columns = {"FIRST_NAME" : "AANTAL"})
sales_staff_temp

POSITION_EN  AANTAL
0                Branch Manager      19
1              District Manager       4
2               General Manager       1
3  Level 1 Sales Representative      13
4  Level 2 Sales Representative      39
5  Level 3 Sales Representative      24
6              Regional Manager       2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [20]:
filtered_sales_staff = sales_staff.groupby("WORK_PHONE", as_index=False)["SALES_STAFF_CODE"].count()
filtered_sales_staff = filtered_sales_staff[filtered_sales_staff["SALES_STAFF_CODE"] > 4]
filtered_sales_staff = filtered_sales_staff.rename(columns = {"SALES_STAFF_CODE" : "Aantal"})
filtered_sales_staff

WORK_PHONE  Aantal
4     +(44) 121 3505267       5
10  +(61) 03 2982  4242       5
11  +31 (0)20 692 93 94       5
13    +33 1 68 94 52 20       5
18     1 (206) 292-0012       5
19     1 (305) 557-4810       6
20     1 (310) 281-5722       5
21     1 (403) 232-5986       5
22     1 (416) 493-5595       5
23     1 (617) 268-6754       6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [23]:
country_site = pd.merge(country, site, left_on = 'COUNTRY_CODE', how = 'left', right_on="COUNTRY_CODE")
nl_only = country_site["COUNTRY"] == 'Netherlands'
country_site.loc[nl_only, ("ADDRESS1", "CITY")]

ADDRESS1          CITY
333                   Beets 36         Beets
334               Perestraat 1      Deventer
335  Laan van Meerdervoort 966      Den Haag
336             Kalfjeslaan 47    Amstelveen
337               Startbaan 11    Amstelveen
338      Noorderspoorsingel 28     Groningen
339              Hogehilweg 13     Amsterdam
340             Spoorstraat 43    Varsseveld
341             Westhavenweg 8     Amsterdam
342              Westzeedijk 4        Muiden
343            Gooiergracht 91         Laren
344            Hoofdstraat 150   Pieterburen
345                Olde Diek 2         Laren
346     Amsterdamsestraatweg 8      Maarssen
347        Krayenhoffstraat 11     Amsterdam
348         Overschiestraat 19     Amsterdam
349              Havenbuurt 12        Marken
350               Duivendaal 1    Wageningen
351              Voorstraat 37  Nieuweschans
352         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [35]:
prod_type = pd.read_sql("SELECT * FROM product_type", con_sales)
prodtype_product = pd.merge(product, prod_type, left_on = "PRODUCT_TYPE_CODE", how = "left", right_on="PRODUCT_TYPE_CODE")
eyewear = prodtype_product["PRODUCT_TYPE_EN"] == "Eyewear"
prodtype_product.loc[eyewear, "PRODUCT_NAME"]

80        Polar Sun
81        Polar Ice
83     Polar Sports
84       Polar Wave
85    Polar Extreme
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [41]:
#pd.merge(sales_staff, sales_branch, left_on= "SALES_BRANCH_CODE", how="left", right_on="SALES_BRANCH_CODE")


Geef een overzicht met daarin van de verkopers hun functie                                                                          
indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [59]:
order_header = pd.read_sql_query("SELECT * FROM order_header", con_sales)
header_staff = pd.merge(sales_staff, order_header, left_on= "SALES_STAFF_CODE", how="left", right_on="SALES_STAFF_CODE")
manager = header_staff["POSITION_EN"].str.contains("Manager")
header_staff.loc[manager, ("POSITION_EN", "ORDER_DATE")].drop_duplicates()


POSITION_EN  ORDER_DATE
421     Branch Manager         NaN
422    General Manager         NaN
1276  Regional Manager         NaN
1643  District Manager         NaN
5325    Branch Manager  2022-09-09
5326    Branch Manager  2022-10-31
5327    Branch Manager  2022-12-13
5328    Branch Manager  2022-03-13
5329    Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [63]:
bigOrder = pd.merge(prodtype_product, order_details, left_on="PRODUCT_NUMBER", how="left", right_on="PRODUCT_NUMBER")
meerdan750 = bigOrder['QUANTITY'] > 750 
bigOrder.loc[meerdan750, ("PRODUCT_NAME", "PRODUCT_TYPE_EN")].drop_duplicates()

PRODUCT_NAME    PRODUCT_TYPE_EN
8616                Star Peg              Tents
15975              Firefly 2           Lanterns
37903        BugShield Spray  Insect Repellents
38313  BugShield Lotion Lite  Insect Repellents
38770       BugShield Lotion  Insect Repellents
39618      BugShield Extreme  Insect Repellents
40754         Sun Shelter 15          Sunscreen
41106         Sun Shelter 30          Sunscreen
41681             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [69]:
bigOrder
korting = ((bigOrder["UNIT_PRICE"] - bigOrder["UNIT_SALE_PRICE"]) / bigOrder["UNIT_PRICE"]) > 0.4
bigOrder.loc[korting, "PRODUCT_NAME"].drop_duplicates()

2765     Hailstorm Titanium Woods Set
37446               BugShield Natural
37967                 BugShield Spray
38306           BugShield Lotion Lite
38772                BugShield Lotion
39561               BugShield Extreme
40799                  Sun Shelter 15
41194                  Sun Shelter 30
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [96]:
returned_item = pd.read_sql_query("SELECT * FROM returned_item", con_sales)
return_reason = pd.read_sql_query("SELECT * FROM return_reason", con_sales)
item_reason = pd.merge(returned_item, return_reason, left_on="RETURN_REASON_CODE", right_on="RETURN_REASON_CODE", how="left")
order_item_reason = pd.merge(item_reason, order_details, left_on="ORDER_DETAIL_CODE", right_on="ORDER_DETAIL_CODE", how="left")

order_item_reason_filter = order_item_reason["RETURN_QUANTITY"].astype(float) / order_item_reason["QUANTITY"].astype(float) > 0.9

filtered_return_descriptions = order_item_reason.loc[order_item_reason_filter, "RETURN_DESCRIPTION_EN"].drop_duplicates()

filtered_return_descriptions


0    Unsatisfactory product
2     Wrong product shipped
3     Wrong product ordered
Name: RETURN_DESCRIPTION_EN, dtype: object

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [100]:
counttype = prodtype_product.groupby("PRODUCT_TYPE_EN", as_index=False)["PRODUCT_NUMBER"].count()
counttype.rename(columns = {"PRODUCT_NUMBER" : "AANTAL"})

PRODUCT_TYPE_EN  AANTAL
0             Binoculars       4
1   Climbing Accessories       7
2           Cooking Gear      10
3                Eyewear       5
4              First Aid       5
5       Golf Accessories       4
6      Insect Repellents       5
7                  Irons       4
8                 Knives       5
9               Lanterns      12
10            Navigation       4
11                 Packs       6
12               Putters       3
13                  Rope       4
14                Safety       4
15         Sleeping Bags       7
16             Sunscreen       5
17                 Tents       6
18                 Tools       6
19               Watches       5
20                 Woods       4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [ ]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [ ]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 